lib/weighted_layers_v2.py

```diff
@@ -0,0 +1,372 @@
+      `virtual_batch_size` is not `None`, instead perform "Ghost Batch
+      Normalization", which creates virtual sub-batches which are each
+      normalized separately (with shared gamma, beta, and moving statistics).
+      Must divide the actual batch size during execution.
+    adjustment: A function taking the `Tensor` containing the (dynamic) shape of
+      the input tensor and returning a pair (scale, bias) to apply to the
+      normalized values (before gamma and beta), only during training. For
+      example, if axis==-1,
+        `adjustment = lambda shape: (
+          tf.random.uniform(shape[-1:], 0.93, 1.07),
+          tf.random.uniform(shape[-1:], -0.1, 0.1))`
+      will scale the normalized value by up to 7% up or down, then shift the
+      result by up to 0.1 (with independent scaling and bias for each feature
+      but shared across all examples), and finally apply gamma and/or beta. If
+      `None`, no adjustment is applied. Cannot be specified if
+      virtual_batch_size is specified.
+  """
+  def __init__(self, num_templates=10, **kwargs):
+    WeightedMixIn.__init__(self)
+    new_kwargs = kwargs.copy()
+    new_kwargs["center"] = False
+    new_kwargs["scale"] = False
+    BatchNormalization.__init__(self, **new_kwargs)
+    self.num_templates = num_templates
+    if "scale" not in kwargs or kwargs["scale"]:
+      self.add_template_variable(weight_name='template_gamma')
+    if "center" not in kwargs or kwargs["center"]:
+      self.add_template_variable(weight_name='template_beta')
+    mixture_input_spec = InputSpec(ndim=1)
+    self.input_spec = (self.input_spec, mixture_input_spec)
+
+  def build(self, input_shape):
+    BatchNormalization.build(self, input_shape)
+    param_shape = (input_shape[self.axis[0]],)
+    self.template_gamma = self.add_weight(
+        name='template_gamma',
+        shape=param_shape,
+        dtype=self._param_dtype,
+        initializer=self.gamma_initializer,
+        regularizer=self.gamma_regularizer,
+        constraint=self.gamma_constraint,
+        trainable=True,
+        experimental_autocast=False)
+    self.template_beta = self.add_weight(
+        name='template_beta',
+        shape=param_shape,
+        dtype=self._param_dtype,
+        initializer=self.beta_initializer,
+        regularizer=self.beta_regularizer,
+        constraint=self.beta_constraint,
+        trainable=True,
+        experimental_autocast=False)
+
+  def call(self, inputs, training=True):
+    layer_inputs = inputs[0]
+    mix_weights = inputs[1]
+    self.assign_mixture_value(name="template_beta", mixture_weights=mix_weights)
+    self.assign_mixture_value(name="template_gamma", mixture_weights=mix_weights)
+    norm_input = BatchNormalization.call(self, layer_inputs, training)
+
+    input_shape = layer_inputs.shape
+    reduction_axes = [i for i in range(len(input_shape))
+                      if i not in self.axis]
+    mean, var = nn.moments(norm_input, reduction_axes, keep_dims=True)
```

Comment:
I assume in the training mode, we really expect mean to be 0 and var to be 1?

Reply:
Yes, but it might be a little bit off these values. The description of BN layer says it will normalize to something close to zero mean and variance 1. I have tried comparing to constants at first, and the corresponding tests crashed.